core/shared/src/main/scala/zio/FiberPool.scala

```diff
@@ -0,0 +1,184 @@
+ * Fiber pools are usually more efficient for long-running processes in which tasks
+ * must be executed in a bounded amount of parallelism compared to repeatedly spawning
+ * new fibers.
+ */
+class FiberPool(state: Ref[State], workerLimit: Long, defectHandler: Cause[Nothing] => UIO[Any]) {
+
+  /**
+   * Submits a new task to the pool. This method will semantically block until a fiber is available
+   * to start the task. Defects caused by the task will be handled by the pool's defect handler.
+   *
+   * If the pool is already shutdown, this method will interrupt its caller.
+   */
+  def submit[R](task: URIO[R, Any]): URIO[R, Any] =
+    ZIO.environment[R].flatMap { r =>
+      Promise.make[Nothing, Unit].flatMap { taskReceived =>
+        val wrappedTask = task.catchAllCause(defectHandler(_).sandbox.ignore).provide(r)
+
+        state.modify {
+          case s @ State(workerCount, freeWorkers, pendingTasks, shutdownType, workers) =>
+            shutdownType match {
+              case None =>
+                if (pendingTasks.nonEmpty) {
+                  // To ensure fairness, we always queue up behind tasks that are
+                  // already pending.
+                  (
+                    taskReceived.await,
+                    State(
+                      workerCount,
+                      freeWorkers,
+                      pendingTasks.enqueue(taskReceived.succeed(()).as(wrappedTask)),
+                      shutdownType,
+                      workers
+                    )
+                  )
+
+                } else if (freeWorkers.nonEmpty) {
+                  // There were no pending tasks, but we did find an existing
+                  // free worker. So hand off the task to it.
+                  val (worker, rest) = freeWorkers.dequeue
+
+                  (worker(wrappedTask), State(workerCount, rest, pendingTasks, shutdownType, workers))
+                } else if (workerCount < workerLimit) {
+                  // No pending tasks, no free workers, but the number of workers
+                  // is below the limit. So launch a new worker.
+                  (addWorker(wrappedTask), State(workerCount + 1, freeWorkers, pendingTasks, shutdownType, workers))
+                } else {
+                  // No free resources, so queue up the task in the pending tasks
+                  // queue.
+                  (
+                    taskReceived.await,
+                    State(
+                      workerCount,
+                      freeWorkers,
+                      pendingTasks.enqueue(taskReceived.succeed(()).as(wrappedTask)),
+                      shutdownType,
+                      workers
+                    )
+                  )
+                }
+              case _ => (ZIO.interrupt, s)
+            }
+        }.flatten
+      }
+
+    }
+
+  /**
+   * Shuts down the fiber pool. Behavior is dependent on the specified [[Shutdown]].
+   */
+  def shutdown(shutdownType: Shutdown): UIO[Any] =
```

Comment:
We should make sure we can replicate this to `Queue#shutdown` so it's not just a one-off data type.

Reply:
On returning `Any`, I think it's best to:

 - Allow users to return `Any` in callbacks, etc.
 - Always return `Unit` ourselves, unless for justifiable performance reasons

Because a `Unit` type is clearer as an output from a ZIO method.